addons/hw_drivers/server_logger.py

```diff
@@ -0,0 +1,152 @@
+        self._active = None
+        self.toggle_active(active)
+
+    def toggle_active(self, is_active):
+        """
+        Switch it on or off the handler (depending on the IoT setting) without the need to close/reset it
+        """
+        self._active = is_active
+        if self._active and self._odoo_server_url:
+            # Start the thread to periodically flush logs
+            self._flush_thread = threading.Thread(target=self._periodic_flush, name="ThreadServerLogSender")
+            self._flush_thread.start()
+        else:
+            self._flush_thread and self._flush_thread.join()  # let a last flush
+
+    def _periodic_flush(self):
+        odoo_session = requests.Session()
+        while self._odoo_server_url and self._active:  # allow to exit the loop on thread.join
+            time.sleep(self._FLUSH_INTERVAL)
+            self._flush_logs(odoo_session)
+
+    def _flush_logs(self, odoo_session):
+        def convert_to_byte(s):
+            return bytes(s, encoding="ascii") + b'<log/>\n'  # Correspond to the "Unit Separator" that I doubt is used
+
+        def convert_server_line(log_level, line_formatted):
+            return convert_to_byte(f"{log_level},{line_formatted}")
+
+        def empty_queue():
+            yield convert_to_byte(f"mac {helpers.get_mac_address()}")
+            for _ in range(self._MAX_BATCH_SIZE):
+                # Use a limit to avoid having too heavy requests & infinite loop of the queue receiving new entries
+                try:
+                    log_record = self._log_queue.get_nowait()
+                    yield convert_server_line(log_record.levelno, self.format(log_record))
+                except queue.Empty:
+                    break
+
+            # Report to the server if the queue is close from saturation
+            if queue_size >= .8 * self._MAX_QUEUE_SIZE:
+                log_message = "The IoT {} queue is saturating: {}/{} ({:.2f}%)".format(
+                    self.__class__.__name__, queue_size, self._MAX_QUEUE_SIZE,
+                    100 * queue_size / self._MAX_QUEUE_SIZE)
+                _logger.warning(log_message)  # As we don't log our own logs, this will be part of the IoT logs
+                # In order to report this to the server (on the current batch) we will append it manually
+                yield convert_server_line(logging.WARNING, log_message)
+
+        queue_size = self._log_queue.qsize()  # This is an approximate value
+
+        if not self._odoo_server_url or queue_size == 0:
+            return
+        try:
+            odoo_session.post(
+                self._odoo_server_url + '/iot/log',
+                data=empty_queue(),
+                timeout=self._REQUEST_TIMEOUT
+            ).raise_for_status()
+        except (requests.exceptions.ConnectTimeout, requests.exceptions.ConnectionError):
+            _logger.info("Connection with the server to send the logs failed. It is likely down")
+        except Exception as _:
```

Comment:
no need for "as _"

Reply:
in theory not indeed, but if we don't the ci/style is not happy

Reply:
Probably because of the catching of the generic "Exception", right? You could avoid this by catching `requests.exceptions.RequestException` instead as the only thing which can throw an exception here is the odoo_session.post and RequestException is the parent class of all the exceptions coming from python requests (see [here](https://requests.readthedocs.io/en/latest/_modules/requests/exceptions/))